### DISCUSSION TITLE
Interactive next/checkout

### DISCUSSION THREAD

**Discussion by bcongdon** [arxanas/git-branchless#138] [Ideas]
I had a couple ideas / feature requests for interactive elements in `git-branchless`, and wanted to start a thread to get feedback before opening PRs:

## Interactive Checkout

It'd be nice to have an interactive checkout command (maybe `pick` or `smartcheckout`?) based on `smartlog`. My initial thought was to display a smartlog with an additional suffix identifier/"hint", followed by a prompt:

Using numeric hints:

<img height="300" alt="Screen Shot 2021-10-19 at 9 10 50 PM" src="https://user-images.githubusercontent.com/706257/138107849-4bd89a3f-8ed5-4c10-9300-0d0c6d04724d.png">

Using [EasyMotion](https://github.com/easymotion/vim-easymotion)/[Vimium](https://github.com/philc/vimium)-style hints:
<img height="300" alt="Screen Shot 2021-10-19 at 9 10 02 PM" src="https://user-images.githubusercontent.com/706257/138107881-bfe6d1df-2b89-43dd-a527-a661c4714cd6.png">

Using [skim](https://crates.io/crates/skim):
<img height="300" alt="Screen Shot 2021-10-19 at 8 48 56 PM" src="https://user-images.githubusercontent.com/706257/138107573-8fb917e1-8a59-4130-9112-56a7be5dff03.png">

@arxanas suggested the EasyMotion/skim approaches. Personally, I prefer the EasyMotion approach here of the three. Skim's UI is quite nice, but it doesn't allow for formatting/coloring of the [fuzzy select item](https://docs.rs/skim/0.9.4/skim/trait.SkimItem.html#required-methods), so it's difficult to nicely display all the information from the smartlog. However, using `skim` would probably be a nicer experience in scenarios where you're picking from a large number of commits.

## Interactive Next

Currently `next` has `--oldest` and `--newest` flags to "guide" the traversal. However, sometimes you have more than 2 commits with the same parent, and an interactive way to pick between them would be nice. Demo interactive selection for `next` (note the `-i` flag):
<img height="300" alt="Screen Shot 2021-10-19 at 8 50 58 PM" src="https://user-images.githubusercontent.com/706257/138107647-91afeacc-62bd-4148-8144-57eaea1f33da.png">

---

Any thoughts?

**Comment by arxanas** [comment]
These are some cool prototypes!

For the command name, I was thinking that it would just be `git branchless checkout`, and then aliased to `git co` by default.

What do you plan to do as far as pagination, when there are many elements in the smartlog, and they don't all fit on the screen?

>  Skim's UI is quite nice, but it doesn't allow for formatting/coloring of the fuzzy select item,

I see that [`display`](https://docs.rs/skim/0.9.4/skim/trait.SkimItem.html#method.display) returns an `AnsiString` — does this not work for that purpose?

Tangentially related: if we use Skim, then it would be nice to highlight the commit element in the preview window, so that the user can get a sense of where in the commit graph they're checking out to. It seems like [`ItemPreview`](https://docs.rs/skim/0.9.4/skim/enum.ItemPreview.html) should also support coloration.

For the interactive next/prev, maybe we should reuse the UI from interactive checkout in some way? That way, the user wouldn't have to get acquainted with multiple ways of selecting commits.
